packages/components/src/ui/context/context-connect.ts

```diff
@@ -16,42 +16,70 @@ import { CONNECT_STATIC_NAMESPACE } from './constants';
 import { getStyledClassNameFromKey } from './get-styled-class-name-from-key';
 import type { WordPressComponentFromProps } from '.';
 
+type AcceptsTwoArgs<
+	F extends ( ...args: any ) => any,
+	ErrorMessage = never
+> = Parameters< F >[ 'length' ] extends 2 ? {} : ErrorMessage;
```

Comment:
Turns out, TypeScript is (justifiably) loose about function arity. No wonder my previous attempts didn't work!

Here is a [minimal repro in a TS Playground](https://www.typescriptlang.org/play?#code/PTAEElQdwQwOwC6gLIE8BiBXOBjUCB7UACxgDcBTfKIgBxgCcYBbAZwFgAoBVWqtLLlABeUAAoYALlBxMzAEYUGAGlDzpshUoCUIgHygyBAJYATANxcuAM2w4ExgnFAAbAgVYV07sdbjSBO10AbwBfK04QUAB5OBdUfBgAawpWUCcqeiZmVXlMJFMCVLgAciQlBgIGUEBeDcAiPa43Dy8fCV1hAzDtCKjACnIUGABzYzxWGEwcKl6uHj5QAEEcSdoEVgAVGnmGQdYAHnRQCgAPBAo4UzSxADobxh3peFR2g0eDUQAFRhYKU4Y99D0AG0Si4zoMEMQSgBdQ4nM4XUAAJlAAH5QGFQBoKJRqpZODY7A4nKBWAhKnBBt4CPtYadzmlArg9L5-KADgAyBZLCgrdabbb-PQhcL4yJgACiDEqfwAhFxSeTKa0YM90aFuqKomtiMY0rrQIN3KZ5WSnEqCGIJLlVV0gA) if you're interested in what's going on here and why it's necessary. (Solution courtesy of [StackOverflow](https://stackoverflow.com/questions/51143582/typescript-is-there-a-way-to-typecheck-function-arity) 😌)